docs/develop-an-agent.md

```diff
@@ -0,0 +1,423 @@
+        -   Hub hostname
+        -   Gateway host name
+        -   Device Id
+        -   Module Id
+        -   Key handle
+-   Build a base64-encoded resource uri
+    -   Using process defined [above](#Step-2-Build-a-bas64-encoded-resource-uri)
+-   Get signed token
+    -   Set REST API URL
+        -   `/sign`
+    -   Set corresponding JSON payload
+        -   JSON payload for /sign:
+            -   Key handle from Get identity call above
+            -   Algorithm: "HMAC-SHA256"
+            -   Parameters:
+                -   Message: "&lt;base64-encoded resource uri&gt;"
+    -   Send request for key service endpoint `/run/aziot/keyd.sock`
+        -   If connection refused, go back to Get identity step above (since IoT Hub device information could have changed)
+    -   On successful response, parse response
+        -   SaS token
+-   Create a `sas_token` following these patterns:
+    -   Device e.g. `"SharedAccessSignature sr=myhubname.azure-devices.net/devices/mydeviceid&sig=hIuFfERqcDBnu84EwVlF01JfiaRvH6A20dMWQW6T4fg%3D&se=1609459200"`
+    -   Module e.g. `"SharedAccessSignature sr=myhubname.azure-devices.net/devices/mydeviceid/modules/mymoduleid&sig=hIuFfERqcDBnu84EwVlF01JfiaRvH6A20dMWQW6T4fg%3D&se=1609459200"`
+    -   Note that `sig` needs to be percent-encoded, but `sr` does not
+    -   Also note that the `SharedAccessSignature` is part of both the parameter name AND the parameter value in the connection string below. So,
+        -   `"SharedAccessSignature=SharedAccessSignature sr=..."`
+-   Build IoT Hub connection string with responses from two API calls above
+    -   `"HostName=<host_name>;DeviceId=<device_id>;SharedAccessSignature=<sas_token>"`
+    -   `"HostName=<host_name>;DeviceId=<device_id>;ModuleId=<module_id>;SharedAccessSignature=<sas_token>"`
+-   Connect to IoT Hub using C SDK API [IoTHubDeviceClient_LL_CreateFromConnectionString](https://docs.microsoft.com/en-us/azure/iot-hub/iot-c-sdk-ref/iothub-device-client-ll-h/iothubdeviceclient-ll-createfromconnectionstring)
+-   After disconnect, repeat all steps starting from Get identity above
+
+## Using the C SDK to connect to IoT Hub for X509 authentication
+
+Currently, host processes using module identities can only authenticate against IoT Hub using SaS authentication method (only method supported by IoT service as of Fe semester). The C SDK team has added support for openssl engines to ensure host processes can perform TLS authentication against IoT Hub. Key Service’s openssl engine will be required by host processes for TLS auth, since X509 private keys are not exposed outside the Key service. It will beneficial for host processes to support device X509 authentication for host processes in the short term (by registering as "device" principals), so that module x509 auth support can be easily added in the future. For code snippet samples for using openssl engines, please reach out to the C SDK team.
```

Comment:
```suggestion
Currently, host processes using module identities can only authenticate against IoT Hub using the SaS authentication method (only method supported by IoT service as of Fe semester). The C SDK team has added support for OpenSSL engines to ensure host processes can perform TLS authentication against IoT Hub. Keys Service’s OpenSSL engine will be required by host processes for TLS authentication, since X.509 private keys are not exposed outside the Keys Service. It is beneficial for host processes to support device X.509 authentication for host processes in the short term (by registering as "device" principals), so that module X.509 authentication support can be easily added in the future. For code snippet samples for using OpenSSL engines, please reach out to the C SDK team.
```

- Do external users know what the `Fe semester` is? Might just want to put a date.